httpx/client.py

```diff
@@ -172,42 +156,62 @@ def merge_headers(
 
         return response
 
-    def _resolve_dispatcher(
+    async def send(
         self,
         request: AsyncRequest,
+        *,
+        stream: bool = False,
         auth: AuthTypes = None,
-        trust_env: bool = False,
         allow_redirects: bool = True,
-    ) -> AsyncDispatcher:
-        dispatcher: AsyncDispatcher = RedirectDispatcher(
-            next_dispatcher=self.dispatch,
-            base_cookies=self.cookies,
-            allow_redirects=allow_redirects,
+        verify: VerifyTypes = None,
+        cert: CertTypes = None,
+        timeout: TimeoutTypes = None,
+        trust_env: bool = None,
+    ) -> AsyncResponse:
+        if request.url.scheme not in ("http", "https"):
+            raise InvalidURL('URL scheme must be "http" or "https".')
+
+        get_response = functools.partial(
+            self.get_response, stream=stream, verify=verify, cert=cert, timeout=timeout
         )
 
-        username: typing.Optional[typing.Union[str, bytes]] = None
-        password: typing.Optional[typing.Union[str, bytes]] = None
-        if auth is None:
-            if request.url.username or request.url.password:
-                username, password = request.url.username, request.url.password
-            elif trust_env:
-                netrc_login = get_netrc_login(request.url.authority)
-                if netrc_login:
-                    username, _, password = netrc_login
-        else:
-            if isinstance(auth, tuple):
-                username, password = auth[0], auth[1]
-            elif callable(auth):
-                dispatcher = CustomAuthDispatcher(
-                    next_dispatcher=dispatcher, auth_callable=auth
-                )
-
-        if username is not None and password is not None:
-            dispatcher = BasicAuthDispatcher(
-                next_dispatcher=dispatcher, username=username, password=password
+        get_response = functools.partial(
+            redirect(allow_redirects, cookies=self.cookies), get_response=get_response
+        )
+
+        auth_middleware = self._get_auth_middleware(
+            request=request,
+            trust_env=self.trust_env if trust_env is None else trust_env,
+            auth=self.auth if auth is None else auth,
+        )
+        if auth_middleware is not None:
+            get_response = functools.partial(auth_middleware, get_response=get_response)
+
+        return await get_response(request)
+
+    def _get_auth_middleware(
+        self, request: AsyncRequest, trust_env: bool, auth: AuthTypes = None
+    ) -> typing.Optional[Middleware]:
+        if isinstance(auth, tuple):
+            return basic_auth(username=auth[0], password=auth[1])
+
+        if callable(auth):
+            return custom_auth(auth=auth)
+
+        assert auth is None
```

Comment:
I noticed we use assertions around the code, I believe it is discouraged since they can be [skipped when running in optimized mode](https://www.quora.com/How-do-I-use-Python-asserts-even-when-optimization-mode-is-used).

Not particularly important in this case though.

Reply:
I guess you're right here — we'd rather want a `TypeError` to be raised or something. I'll update the PR!